odl/space/base_ntuples.py

```diff
@@ -403,6 +403,281 @@ def __repr__(self):
+        Numpy ufuncs`_.
+
+        .. note::
+            This basic implementation casts inputs and
+            outputs to Numpy arrays and evaluates ``ufunc`` on those.
+            For `numpy.ndarray` based data storage, this incurs no
+            significant overhead compared to direct usage of Numpy arrays.
+
+            For other (in particular non-local) implementations, e.g.,
+            GPU arrays or distributed memory, overhead is significant due
+            to copies to CPU main memory. In those classes, the
+            ``__array_ufunc__`` mechanism should be overridden to use
+            native implementations if possible.
+
+        .. note::
+            When using operations that alter the shape (like ``reduce``),
+            or the data type (can be any of the methods),
+            the resulting array is wrapped in a space of the same
+            type as ``self.space``, however **only** using the minimal
+            set of parameters ``size`` and ``dtype``. If more properties
+            are supposed to be propagated, this method must be overridden.
+
+        Parameters
+        ----------
+        ufunc : `numpy.ufunc`
+            Ufunc that should be called on ``self``.
+        method : str
+            Method on ``ufunc`` that should be called on ``self``.
+            Possible values:
+
+            ``'__call__'``, ``'accumulate'``, ``'at'``, ``'outer'``,
+            ``'reduce'``, ``'reduceat'``.
+
+        input1, ..., inputN:
+            Positional arguments to ``ufunc.method``.
+        kwargs:
+            Keyword arguments to ``ufunc.method``.
+
+        Returns
+        -------
+        ufunc_result : # TODO fill in
+            Result of the ufunc evaluation.
+
+        Examples
+        --------
+        We apply `numpy.add` to ODL vectors, also with the optional
+        ``out`` parameter:
+
+        >>> rn = odl.rn(3)
+        >>> x = rn.element([1, 2, 3])
+        >>> y = rn.element([-1, -2, -3])
+        >>> out = rn.element()
+        >>> np.add(x, y)
+        rn(3).element([0.0, 0.0, 0.0])
+        >>> result = np.add(x, y, out=out)
+        >>> out
+        rn(3).element([0.0, 0.0, 0.0])
+        >>> result is out
+        True
+
+        The ``add.accumulate`` method retains the original shape and
+        ``dtype``. We can override the latter with the ``dtype``
+        parameter:
+
+        >>> x = rn.element([1, 2, 3])
+        >>> np.add.accumulate(x)
+        rn(3).element([1.0, 3.0, 6.0])
+        >>> np.add.accumulate(x, dtype=complex)
+        cn(3).element([(1+0j), (3+0j), (6+0j)])
+
+        The ``add.at`` method operates in-place. Here we add the second
+        operand ``[5, 10]`` to ``x`` at indices ``[0, 2]``:
+
+        >>> x = rn.element([1, 2, 3])
+        >>> np.add.at(x, [0, 2], [5, 10])
+        >>> x
+        rn(3).element([6.0, 2.0, 13.0])
+
+        The output of ``add.outer`` would have ``ndim = 2`` axes, so we
```

Comment:
Well, since there is no acute use-case of this, why don't we simply leave it undefined for now and then later change it to work like it does for arrays. Having wildly different behavior is just weird imo.

Also we'll want stuff like `np.multiply.outer(odl.uniform_discr(0, 1, 10).one(), [1, 2])` to work decently, I'd personally expect it to return a vector valued space.